### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)


## Player Count

* Display the total number of players


In [2]:
total=purchase_data["SN"].nunique()
total_players= {"total Players":[total]}
df=pd.DataFrame(total_players,index=[0])
df


total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [8]:
#number of unique items
purchase_data_groubby= purchase_data.groupby("Item ID")
unique_item=len(purchase_data_groubby["Item ID"].nunique())

#average price(I could not round a single value)
average_price=purchase_data["Price"].mean()

#number of perchases
number_purchases=purchase_data["Item ID"].count()

#Total Revenue
Total_Revenue=purchase_data["Price"].sum()

#creat a datafram to hold the results
Purchasing_Analysis=pd.DataFrame({"Number of Unique Items":unique_item,
                                 "Average Price":[average_price],
                                 "Number of Purchases":number_purchases,
                                  "Total Revenue":[Total_Revenue]},index=[0])

#give the displayed data cleaner formatting
Purchasing_Analysis["Average Price"]=Purchasing_Analysis["Average Price"].map("${:,.2f}".format)

Purchasing_Analysis["Total Revenue"]=Purchasing_Analysis["Total Revenue"].map("${:,.2f}".format)

#Display the datafram
Purchasing_Analysis

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     183         $3.05                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [7]:
#count of players
total_count_groubby_gender=purchase_data.groupby(["Gender"]).nunique()

#percentage of players
total_count_groubby_gender_percentage=total_count_groubby_gender["SN"]/total*100

#give the displayed data cleaner formatting
total_count_groubby_gender_percentage=total_count_groubby_gender_percentage.map("{:,.2f}%".format)

#create a dataframe to hold the results
Gender_Demographics=pd.DataFrame({"Total Count":total_count_groubby_gender["SN"],
                           "Percentage of Players":total_count_groubby_gender_percentage},
                                  index=["Male", "Female", "Other / Non-Disclosed"])

#display the results
Gender_Demographics

Total Count Percentage of Players
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [11]:
#purchase count
pd_group_gender=purchase_data.groupby(["Gender"])
total_count_2=pd_group_gender["SN"].count()

#average price
ave_count=pd_group_gender["Price"].mean()
ave_count=ave_count.map("{:,.2f}%".format)

#Total Purchase Value(T_P_V)
T_P_V=pd_group_gender["Price"].sum()

#Avg Total Purchase per Person(A_T_P_P_P)
A_T_P_P_P=T_P_V/Gender_Demographics["Total Count"]

#clear formating
T_P_V=T_P_V.map("${:,.2f}".format)
A_T_P_P_P=A_T_P_P_P.map("${:,.2f}".format)

#Create a summary data frame to hold the results 
summery_gender=pd.DataFrame({"Total Count":total_count_2,
                           "Average Price":ave_count,
                        "Total Purchase Value":T_P_V,
                          "Avg Total Purchase Per Person": A_T_P_P_P 
                        })
summery_gender


Total Count Average Price Total Purchase Value  \
Gender                                                                  
Female                         113         3.20%              $361.94   
Male                           652         3.02%            $1,967.64   
Other / Non-Disclosed           15         3.35%               $50.19   

                      Avg Total Purchase Per Person  
Gender                                               
Female                                        $4.47  
Male                                          $4.07  
Other / Non-Disclosed                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [12]:
#Age Demographics
bins=[0,9,14,19,24,29,34,39,2000]
labels_group=["<10","10-14","15-19","20-24","25-29","30-34","35-39","40+"]
purchase_data["score summery"]=pd.cut(purchase_data["Age"],bins, labels=labels_group)

age_group=purchase_data.groupby(["score summery"])

#Calculate the numbers and percentages by age group(n_age_groups) and (p_age_groups)
n_age_groups=age_group["SN"].nunique()
p_age_groups=n_age_groups/total*100

# round the percentage column to two decimal points
p_age_groups=p_age_groups.map("{:,.2f}%".format)


#Create a summary data frame to hold the results 
Age_Demographics=pd.DataFrame({"Total Count":n_age_groups,
                           "Percentage of Players":p_age_groups,
                       
                        })
Age_Demographics



Total Count Percentage of Players
score summery                                   
<10                     17                 2.95%
10-14                   22                 3.82%
15-19                  107                18.58%
20-24                  258                44.79%
25-29                   77                13.37%
30-34                   52                 9.03%
35-39                   31                 5.38%
40+                     12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [14]:
#purchase count
total_count_age=age_group["Purchase ID"].nunique()

#Average Purchase Price(A_P_P_A)
A_P_P_A=age_group["Price"].mean()

#Total purchase value (abreviation:tpv)
tpv=total_count_age*A_P_P_A

#Avg Total Purchase per Person(abreviation:atpp),total count including repetetive customer(abreviation:tcirc)
tcirc=age_group["Purchase ID"].count()
atpp =tpv/Age_Demographics["Total Count"]

#give the displayed data cleaner formatting
A_P_P_A=A_P_P_A.map("${:,.2f}".format)
tpv=tpv.map("${:,.2f}".format)
atpp=atpp.map("${:,.2f}".format)

#Create a summary data frame to hold the results
Purchasing_Analysis_Age=pd.DataFrame({"purchase count":total_count_age,
                                     "Average Purchase Price":A_P_P_A,
                                      "Total purchase value":tpv,
                                      "Avg Total Purchase per Person":atpp                                   
                                     
                                     })
                        
Purchasing_Analysis_Age


purchase count Average Purchase Price Total purchase value  \
score summery                                                               
<10                        23                  $3.35               $77.13   
10-14                      28                  $2.96               $82.78   
15-19                     136                  $3.04              $412.89   
20-24                     365                  $3.05            $1,114.06   
25-29                     101                  $2.90              $293.00   
30-34                      73                  $2.93              $214.00   
35-39                      41                  $3.60              $147.67   
40+                        13                  $2.94               $38.24   

              Avg Total Purchase per Person  
score summery                                
<10                                   $4.54  
10-14                                 $3.76  
15-19                                 $3.86  
20-24                                 $4.32  
25-29                                 $3.81  
30-34                                 $4.12  
35-39                                 $4.76  
40+                                   $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [15]:
#number of unique items(abreviation:nui)
#Group by Spendors ( "SN" )
nui=purchase_data.groupby(["SN"])
nui_count=nui["Purchase ID"].count()


#Average Purchase Price
nui_ave=nui["Price"].mean()


#Total Purchase Value
nui_sum=nui["Price"].sum()

#give the displayed data cleaner formatting
nui_ave=nui_ave.map("${:,.2f}".format)


#Create a summary data frame to hold the results
Top_Spenders=pd.DataFrame({"purchase count":nui_count,
                                     "Average Purchase Price":nui_ave,
                                      "Total purchase value":nui_sum,
                                     
                                     })
#Sort the total purchase value column in descending order

Top_Spenders=Top_Spenders.sort_values("Total purchase value",ascending=False)


#give the displayed data cleaner formatting
Top_Spenders["Total purchase value"]=nui_sum.map("${:,.2f}".format)


Top_Spenders

purchase count Average Purchase Price Total purchase value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10
...                     ...                    ...                  ...
Ililsasya43               1                  $1.02                $1.02
Irilis75                  1                  $1.02                $1.02
Aidai61                   1                  $1.01                $1.01
Chanirra79                1                  $1.01                $1.01
Alo38                     1                  $1.00                $1.00

[576 rows x 3 columns]

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [17]:
#Group by Item ID and Item Name(abbreviation:giiin)
giiin=purchase_data.groupby(["Item ID","Item Name"])

#Purchase Count(abbreviation:pc)
pc=giiin["Purchase ID"].nunique()

#Item Price	(abbreviasion:ip)
ip=giiin["Price"].mean()


#total purchase value by item(tpvi)
tpvi=giiin["Price"].sum()

#item price formating
ip=ip.map("${:,.2f}".format)
tpvi=tpvi.map("${:,.2f}".format)

#Create a summary data frame to hold the results
Most_Popular_Items=pd.DataFrame({"purchase count":pc,
                                     "Item Price":ip,
                                      "Total purchase value":tpvi,     
                                })
#Sort the purchase count in descending order
Most_Popular_Items=Most_Popular_Items.sort_values("purchase count",ascending=False)


Most_Popular_Items

purchase count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
108     Extraction, Quickblade Of Trembling Hands                  9   
82      Nirvana                                                    9   
19      Pursuit, Cudgel of Necromancy                              8   
...                                                              ...   
104     Gladiator's Glaive                                         1   
23      Crucifer                                                   1   
180     Stormcaller                                                1   
91      Celeste                                                    1   
134     Undead Crusader                                            1   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
108     Extraction, Quickblade Of Trembling Hands         $3.53   
82      Nirvana                                           $4.90   
19      Pursuit, Cudgel of Necromancy                     $1.02   
...                                                         ...   
104     Gladiator's Glaive                                $1.93   
23      Crucifer                                          $1.99   
180     Stormcaller                                       $3.36   
91      Celeste                                           $4.17   
134     Undead Crusader                                   $4.50   

                                                     Total purchase value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
82      Nirvana                                                    $44.10  
19      Pursuit, Cudgel of Necromancy                               $8.16  
...                                                                   ...  
104     Gladiator's Glaive                                          $1.93  
23      Crucifer                                                    $1.99  
180     Stormcaller                                                 $3.36  
91      Celeste                                                     $4.17  
134     Undead Crusader                                             $4.50  

[183 rows x 3 columns]

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [20]:
#total purchase value by item(tpvi)
tpvi=giiin["Price"].sum()


#Create a summary data frame to hold the results
Most_Profitable_Items=pd.DataFrame({"purchase count":pc,
                                    "Item Price":ip,
                                     "Total purchase value":tpvi,     
                               })
#Sort the purchase count in descending order
Most_Profitable_Items=Most_Profitable_Items.sort_values("Total purchase value",ascending=False)
Most_Profitable_Items["Total purchase value"]=Most_Profitable_Items["Total purchase value"].map("${:,.2f}".format)

Most_Profitable_Items

purchase count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
92      Final Critic                                               8   
103     Singed Scalpel                                             8   
...                                                              ...   
125     Whistling Mithril Warblade                                 2   
126     Exiled Mithril Longsword                                   1   
23      Crucifer                                                   1   
104     Gladiator's Glaive                                         1   
42      The Decapitator                                            1   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
92      Final Critic                                      $4.88   
103     Singed Scalpel                                    $4.35   
...                                                         ...   
125     Whistling Mithril Warblade                        $1.00   
126     Exiled Mithril Longsword                          $2.00   
23      Crucifer                                          $1.99   
104     Gladiator's Glaive                                $1.93   
42      The Decapitator                                   $1.75   

                                                     Total purchase value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
92      Final Critic                                               $39.04  
103     Singed Scalpel                                             $34.80  
...                                                                   ...  
125     Whistling Mithril Warblade                                  $2.00  
126     Exiled Mithril Longsword                                    $2.00  
23      Crucifer                                                    $1.99  
104     Gladiator's Glaive                                          $1.93  
42      The Decapitator                                             $1.75  

[183 rows x 3 columns]

Conclusion

-The total revenue from 183 items was $2,379.77. 

-Among 576 unique and 780 non-unique players, more than 84.03% was male and more than 14.06% was female. 

-Based on purchasing analysis (Gender), although the number of female players was much lower than male players, the average of money spend by  female group was a little higher than the male group ($4.47 vs $4.07).

-Based on the age demographics results, 44.79% of customers were in 20-24 age groups which represented the most interested groups. Then the age group between 15-19-year-old with 18.58% took the second position. 

- Based on purchasing analysis (Age), total purchase value in the age range between 20-24 was the highest, while in the category of 35-39 the average total per person was the highest.

-Based on top spenders results, the best customers were Lisosia93, Idastidru52 and Chamjask73 which spend $18.96, $15.45 and $13.83, respectively. Average purchase is about $3 per person.

-Based on most popular and most profitable items results, the most popular and profitable items was “Oathbreaker, Last Hope of the Breaking Storm” with total purchase of $50.76 (12 buyes). After this, the second and third most favorite items were Nirvana" and "Fiery Glass Crusader"  which brough $44.10 and $41.22 respectively (9 buys).






